src/lib/optimizer/chunk_statistics/chunk_column_statistics.cpp

```diff
@@ -0,0 +1,88 @@
+#include "chunk_column_statistics.hpp"
+
+#include <iterator>
+
+#include "resolve_type.hpp"
+
+#include "optimizer/chunk_statistics/abstract_filter.hpp"
+#include "optimizer/chunk_statistics/min_max_filter.hpp"
+#include "optimizer/chunk_statistics/range_filter.hpp"
+#include "storage/base_encoded_column.hpp"
+#include "storage/deprecated_dictionary_column.hpp"
+#include "storage/dictionary_column.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/run_length_column.hpp"
+#include "storage/value_column.hpp"
+
+namespace opossum {
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_dictionary(const pmr_vector<T>& dictionary) {
+  auto statistics = std::make_shared<ChunkColumnStatistics>();
+  // only create statistics when the compressed dictionary is not empty
+  if (!dictionary.empty()) {
+    auto min_max_filter = std::make_unique<MinMaxFilter<T>>(dictionary.front(), dictionary.back());
+    statistics->add_filter(std::move(min_max_filter));
+
+    // no range filter for strings
+    // clang-format off
+    if constexpr(std::is_arithmetic_v<T>) {
+      auto range_filter = RangeFilter<T>::build_filter(dictionary);
+      statistics->add_filter(std::move(range_filter));
+    }
+    // clang-format on
+  }
+  return statistics;
+}
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const DictionaryColumn<T>& column) {
+  const auto& dictionary = *column.dictionary();
+  return build_statistics_from_dictionary(dictionary);
+}
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ValueColumn<T>& column) {
+  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
+  return std::make_shared<ChunkColumnStatistics>();
+}
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(
+    const DeprecatedDictionaryColumn<T>& column) {
+  const auto& dictionary = *column.dictionary();
+  return build_statistics_from_dictionary(dictionary);
+}
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ReferenceColumn& column) {
+  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
+  return std::make_shared<ChunkColumnStatistics>();
+}
+
+template <typename T>
+static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const RunLengthColumn<T>& column) {
+  // TODO(tbjoern): implement build_statistics_from_concrete_column
+  return std::make_shared<ChunkColumnStatistics>();
```

Comment:
I’m still not happy with how the different column types are handled here. For one, it requires adding specialisations for every new encoding scheme. That’s why I proposed to replace it with a more general specialisation that would work for all column types.

```c++
static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const BaseColumn& column) {
  Fail("Error message");
}
```

Simply replacing the other functions with this one doesn’t work because the calling code still passes a template argument; the new function however doesn’t accept one. The functions accepting dictionary columns do not need an explicit template argument because it can be deduced from their parameters. So all you have to do is remove the template argument in line 74, and it should work. At least when I tested it it did. Line 73 won’t be needed anymore either.

```c++
statistics = build_statistics_from_concrete_column(typed_column);
```

Reply:
You dont want this method to fail for new encoding schemes. It has to be implemented differently, depending on the encoding scheme. If there is a new encoding scheme, default behavior should be to create no filters at all, until someone implements this build method for the new scheme.

Reply:
Okay then keep the function for reference columns and return a nullptr for in function that I proposed. Currently, the code wouldn’t compile without the addition of new function.

Reply:
I disagree, @tbjoern . The existence of the statistics should not depend on the encoding scheme. Why should they?